### PR TITLE
refactor: abstract `std::fs` away from on-disk index cache

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

Part of https://github.com/rust-lang/cargo/issues/6908

This abstracts `std::fs` so that we can integrate SQLite into on-disk cache in follow-ups.

### How should we test and review this PR?

* Make sure the documentation still render well.
* Shouldn't have any behavior change. Only a bit of extra allocations added.

### CODE REVIEW

File: src/cargo/sources/registry/index/cache.rs [hunk 1]
```diff
@@ -212,3 +218,48 @@ impl<'a> SummariesCache<'a> {
         contents
     }
 }
+
+/// Manages the on-disk index caches.
+pub struct CacheManager<'gctx> {
+    /// [`GlobalContext`] reference for convenience.
+    pub gctx: &'gctx GlobalContext,
```
Comment (epage) [inline src/cargo/sources/registry/index/cache.rs:228]: Feels weird to me that callers access the context through this

Comment (weihanglo) [reply src/cargo/sources/registry/index/cache.rs]: Fixed!

Comment (epage) [reply src/cargo/sources/registry/index/cache.rs]: Passing in individual unstable features seems a bit higher churn as we evolve but whichever way.

**Pull request merged**